replication.go

```diff
@@ -0,0 +1,288 @@
+	// used for lag tracking
+	ServerWalEnd uint64
+	// The server time, in microseconds since jan 1 2000
+	ServerTime uint64
+	// If 1, the server is requesting a standby status message
+	// to be sent immediately.
+	ReplyRequested byte
+}
+
+func (s *ServerHeartbeat) Time() time.Time {
+	return time.Unix(0, (int64(s.ServerTime)*1000)+epochNano)
+}
+
+func (s *ServerHeartbeat) String() string {
+	return fmt.Sprintf("WalEnd: %s ReplyRequested: %d T: %s", FormatLSN(s.ServerWalEnd), s.ReplyRequested, s.Time())
+}
+
+// The replication message wraps all possible messages from the
+// server received during replication. At most one of the wal message
+// or server heartbeat will be non-nil
+type ReplicationMessage struct {
+	WalMessage      *WalMessage
+	ServerHeartbeat *ServerHeartbeat
+}
+
+// The standby status is the client side heartbeat sent to the postgresql
+// server to track the client wal positions. For practical purposes,
+// all wal positions are typically set to the same value.
+type StandbyStatus struct {
+	// The WAL position that's been locally written
+	WalWritePosition uint64
+	// The WAL position that's been locally flushed
+	WalFlushPosition uint64
+	// The WAL position that's been locally applied
+	WalApplyPosition uint64
+	// The client time in microseconds since jan 1 2000
+	ClientTime uint64
+	// If 1, requests the server to immediately send a
+	// server heartbeat
+	ReplyRequested byte
+}
+
+// Create a standby status struct, which sets all the WAL positions
+// to the given wal position, and the client time to the current time.
+func NewStandbyStatus(walPosition uint64) (status *StandbyStatus) {
+	status = new(StandbyStatus)
+	status.WalFlushPosition = walPosition
+	status.WalApplyPosition = walPosition
+	status.WalWritePosition = walPosition
+	status.ClientTime = uint64((time.Now().UnixNano() - epochNano) / 1000)
+	return
+}
+
+// Send standby status to the server, which both acts as a keepalive
+// message to the server, as well as carries the WAL position of the
+// client, which then updates the server's replication slot position.
+func (c *Conn) SendStandbyStatus(k *StandbyStatus) (err error) {
+	writeBuf := newWriteBuf(c, copyData)
+	writeBuf.WriteByte(standbyStatusUpdate)
+	writeBuf.WriteInt64(int64(k.WalWritePosition))
+	writeBuf.WriteInt64(int64(k.WalFlushPosition))
+	writeBuf.WriteInt64(int64(k.WalApplyPosition))
+	writeBuf.WriteInt64(int64(k.ClientTime))
+	writeBuf.WriteByte(k.ReplyRequested)
+
+	writeBuf.closeMsg()
+
+	_, err = c.conn.Write(writeBuf.buf)
+	if err != nil {
+		c.die(err)
+	}
+
+	return
+}
+
+// Send the message to formally stop the replication stream. This
+// is done before calling Close() during a clean shutdown.
+func (c *Conn) StopReplication() (err error) {
+	writeBuf := newWriteBuf(c, copyDone)
+
+	writeBuf.closeMsg()
+
+	_, err = c.conn.Write(writeBuf.buf)
+	if err != nil {
+		c.die(err)
+	}
+	return
+}
+
+
+func (c *Conn) readReplicationMessage() (r *ReplicationMessage, err error) {
```

Comment:
Do you want to handle a command byte that's not defined with an error message or just ignore it?